File: packages/core/src/__tests__/git.test.ts
/**
 * Test suite for GitWrapper with isomorphic-git
 */

import { GitWrapper } from '../git';
import { FileSystem } from '../file-system';
import * as path from 'path';
import * as os from 'os';

describe('GitWrapper', () => {
  let tempDir: string;
  let gitWrapper: GitWrapper;
  let testFile: string;

  beforeEach(async () => {
    tempDir = path.join(os.tmpdir(), `git-test-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`);
    await FileSystem.mkdir(tempDir);
    
    gitWrapper = new GitWrapper(tempDir);
    testFile = path.join(tempDir, 'test.txt');
    await FileSystem.writeFile(testFile, 'Hello World');
  });

  afterEach(async () => {
    try {
      const fs = await import('fs/promises');
      await fs.rm(tempDir, { recursive: true, force: true });
    } catch {
      // Ignore cleanup errors
    }
  });

  describe('Git Initialization', () => {
    test('should initialize git repository', async () => {
      await gitWrapper.init();
      
      const gitDir = path.join(tempDir, '.git');
      expect(await FileSystem.exists(gitDir)).toBe(true);
    });

    test('should check if repository is healthy after init', async () => {
      await gitWrapper.init();
      expect(await gitWrapper.isHealthy()).toBe(true);
    });

    test('should return false for health check before init', async () => {
      expect(await gitWrapper.isHealthy()).toBe(false);
    });
  });

  describe('Git Operations', () => {
    beforeEach(async () => {
      await gitWrapper.init();
    });

    test('should add files to git', async () => {
      await expect(gitWrapper.add(testFile)).resolves.not.toThrow();
    });

    test('should commit changes', async () => {
      await gitWrapper.add(testFile);
      await expect(gitWrapper.commit('Initial commit')).resolves.not.toThrow();
    });

    test('should get status', async () => {
      const status = await gitWrapper.status();
      expect(status).toBeDefined();
      expect(status.files).toBeDefined();
    });

    test('should generate diff for modified file', async () => {
      // Add and commit initial file
      await gitWrapper.add(testFile);
      await gitWrapper.commit('Initial commit');
      
      // Modify file
      await FileSystem.writeFile(testFile, 'Hello Modified World');
      
      const diff = await gitWrapper.diff(testFile);
      expect(diff).toBeDefined();
    });

    test('should show no diff for unmodified file', async () => {
      // Add and commit file
      await gitWrapper.add(testFile);
      await gitWrapper.commit('Initial commit');
      
      const diff = await gitWrapper.diff(testFile);
      expect(diff).toBe('');
    });

    test('should reset file to HEAD', async () => {
      // Add and commit file
      await gitWrapper.add(testFile);
      await gitWrapper.commit('Initial commit');
      
      await expect(gitWrapper.reset(testFile)).resolves.not.toThrow();
    });

    test('should checkout file from HEAD', async () => {
      // Add and commit file
      await gitWrapper.add(testFile);
      await gitWrapper.commit('Initial commit');
      
      // Modify file
      await FileSystem.writeFile(testFile, 'Modified content');
      
      // Checkout original version
      await gitWrapper.checkout(testFile);
      
      const content = await FileSystem.readFile(testFile);
      expect(content).toBe('Hello World');
    });
  });

  describe('Git History', () => {
    beforeEach(async () => {
      await gitWrapper.init();
    });

    test('should have no commits initially', async () => {
      expect(await gitWrapper.hasCommits()).toBe(false);
      expect(await gitWrapper.getCommitCount()).toBe(0);
    });

    test('should count commits correctly', async () => {
      await gitWrapper.add(testFile);
      await gitWrapper.commit('First commit');
      
      expect(await gitWrapper.hasCommits()).toBe(true);
      expect(await gitWrapper.getCommitCount()).toBe(1);
      
      // Add another commit
      const file2 = path.join(tempDir, 'test2.txt');
      await FileSystem.writeFile(file2, 'Second file');
      await gitWrapper.add(file2);
      await gitWrapper.commit('Second commit');
      
      expect(await gitWrapper.getCommitCount()).toBe(2);
    });
  });

  describe('Error Handling', () => {
    test('should handle git operations before init gracefully', async () => {
      // Most operations should auto-initialize, but let's test explicit behavior
      await expect(gitWrapper.add(testFile)).resolves.not.toThrow();
    });

    test('should handle non-existent files', async () => {
      await gitWrapper.init();
      
      const nonExistentFile = path.join(tempDir, 'nonexistent.txt');
      await expect(gitWrapper.add(nonExistentFile)).rejects.toThrow();
    });

    test('should handle checkout of non-existent file', async () => {
      await gitWrapper.init();
      
      const nonExistentFile = path.join(tempDir, 'nonexistent.txt');
      await expect(gitWrapper.checkout(nonExistentFile)).rejects.toThrow();
    });

    test('should handle diff of uncommitted file', async () => {
      await gitWrapper.init();
      
      const diff = await gitWrapper.diff(testFile);
      expect(diff).toContain('+');
    });
  });

  describe('Git Configuration', () => {
    test('should set up git config during init', async () => {
      await gitWrapper.init();
      
      // Git should be configured with Oops user
      // We can't easily test config values with isomorphic-git, but init should not throw
      expect(await gitWrapper.isHealthy()).toBe(true);
    });
  });
});

File: internal/git/git_test.go
package git

import (
	"os"
	"path/filepath"
	"testing"
)

func setupTestRepo(t *testing.T) (*Repo, string, func()) {
	tmpDir := t.TempDir()
	testFile := "test.txt"
	testFilePath := filepath.Join(tmpDir, testFile)

	// Create test file
	if err := os.WriteFile(testFilePath, []byte("initial content"), 0644); err != nil {
		t.Fatal(err)
	}

	gitDir := filepath.Join(tmpDir, ".oops", testFile+".git")
	repo := NewRepo(gitDir, tmpDir, testFile)

	cleanup := func() {}
	return repo, tmpDir, cleanup
}

func TestRepoInit(t *testing.T) {
	repo, _, cleanup := setupTestRepo(t)
	defer cleanup()

	if err := repo.Init(); err != nil {
		t.Fatalf("Init failed: %v", err)
	}

	if !repo.Exists() {
		t.Error("Repo should exist after Init")
	}
}

func TestRepoInitAndCommit(t *testing.T) {
	repo, tmpDir, cleanup := setupTestRepo(t)
	defer cleanup()

	if err := repo.Init(); err != nil {
		t.Fatal(err)
	}

	// Add and commit
	if err := repo.Add(); err != nil {
		t.Fatalf("Add failed: %v", err)
	}

	_, err := repo.Commit("Initial commit")
	if err != nil {
		t.Fatalf("Commit failed: %v", err)
	}

	// Tag
	if err := repo.Tag("v1"); err != nil {
		t.Fatalf("Tag failed: %v", err)
	}

	// Check latest tag
	num, err := repo.GetLatestTagNumber()
	if err != nil {
		t.Fatalf("GetLatestTagNumber failed: %v", err)
	}
	if num != 1 {
		t.Errorf("Latest tag = %d, want 1", num)
	}

	// Modify and commit again
	testFilePath := filepath.Join(tmpDir, "test.txt")
	if err := os.WriteFile(testFilePath, []byte("modified content"), 0644); err != nil {
		t.Fatal(err)
	}

	if err := repo.Add(); err != nil {
		t.Fatal(err)
	}

	_, err = repo.Commit("Second commit")
	if err != nil {
		t.Fatal(err)
	}

	if err := repo.Tag("v2"); err != nil {
		t.Fatal(err)
	}

	num, _ = repo.GetLatestTagNumber()
	if num != 2 {
		t.Errorf("Latest tag = %d, want 2", num)
	}
}

func TestRepoCheckout(t *testing.T) {
	repo, tmpDir, cleanup := setupTestRepo(t)
	defer cleanup()

	testFilePath := filepath.Join(tmpDir, "test.txt")

	// Init and first commit
	repo.Init()
	repo.Add()
	repo.Commit("v1")
	repo.Tag("v1")

	// Second version
	os.WriteFile(testFilePath, []byte("version 2"), 0644)
	repo.Add()
	repo.Commit("v2")
	repo.Tag("v2")

	// Checkout v1
	if err := repo.Checkout("v1"); err != nil {
		t.Fatalf("Checkout failed: %v", err)
	}

	content, _ := os.ReadFile(testFilePath)
	if string(content) != "initial content" {
		t.Errorf("Content = %q, want %q", string(content), "initial content")
	}
}

func TestRepoHasChanges(t *testing.T) {
	repo, tmpDir, cleanup := setupTestRepo(t)
	defer cleanup()

	testFilePath := filepath.Join(tmpDir, "test.txt")

	repo.Init()
	repo.Add()
	repo.Commit("Initial")
	repo.Tag("v1")

	// No changes initially
	hasChanges, err := repo.HasChanges()
	if err != nil {
		t.Fatal(err)
	}
	if hasChanges {
		t.Error("Should not have changes initially")
	}

	// Modify file
	os.WriteFile(testFilePath, []byte("modified"), 0644)

	hasChanges, err = repo.HasChanges()
	if err != nil {
		t.Fatal(err)
	}
	if !hasChanges {
		t.Error("Should have changes after modification")
	}
}

func TestRepoDiff(t *testing.T) {
	repo, tmpDir, cleanup := setupTestRepo(t)
	defer cleanup()

	testFilePath := filepath.Join(tmpDir, "test.txt")

	repo.Init()
	repo.Add()
	repo.Commit("Initial")
	repo.Tag("v1")

	// Modify file
	os.WriteFile(testFilePath, []byte("modified content"), 0644)

	diff, err := repo.Diff()
	if err != nil {
		t.Fatalf("Diff failed: %v", err)
	}

	if diff == "" {
		t.Error("Diff should not be empty")
	}
}

func TestRepoLog(t *testing.T) {
	repo, tmpDir, cleanup := setupTestRepo(t)
	defer cleanup()

	testFilePath := filepath.Join(tmpDir, "test.txt")

	repo.Init()
	repo.Add()
	repo.Commit("First commit")
	repo.Tag("v1")

	os.WriteFile(testFilePath, []byte("v2"), 0644)
	repo.Add()
	repo.Commit("Second commit")
	repo.Tag("v2")

	snapshots, err := repo.Log()
	if err != nil {
		t.Fatalf("Log failed: %v", err)
	}

	if len(snapshots) < 2 {
		t.Errorf("Expected at least 2 snapshots, got %d", len(snapshots))
	}
}

func TestRepoExists(t *testing.T) {
	repo, _, cleanup := setupTestRepo(t)
	defer cleanup()

	if repo.Exists() {
		t.Error("Repo should not exist before Init")
	}

	repo.Init()

	if !repo.Exists() {
		t.Error("Repo should exist after Init")
	}
}

func TestRepoCheckoutHead(t *testing.T) {
	repo, tmpDir, cleanup := setupTestRepo(t)
	defer cleanup()

	testFilePath := filepath.Join(tmpDir, "test.txt")

	repo.Init()
	repo.Add()
	repo.Commit("Initial")
	repo.Tag("v1")

	// Modify file
	os.WriteFile(testFilePath, []byte("unsaved changes"), 0644)

	// Checkout HEAD (undo changes)
	if err := repo.CheckoutHead(); err != nil {
		t.Fatalf("CheckoutHead failed: %v", err)
	}

	content, _ := os.ReadFile(testFilePath)
	if string(content) != "initial content" {
		t.Errorf("Content = %q, want %q", string(content), "initial content")
	}
}
